File: src/main/java/com/jackmeng/stl/stl_EventQueue.java
// Copyright 2023 Jack Meng. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package com.jackmeng.stl;

import java.util.LinkedList;

public class stl_EventQueue
    implements
    Runnable
{

  private final LinkedList< Runnable > eventQueue;
  private boolean isProcessingEvents;

  public stl_EventQueue()
  {
    eventQueue = new LinkedList<>();
    isProcessingEvents = false;
  }

  public synchronized void add(Runnable event)
  {
    eventQueue.add(event);
    if (!isProcessingEvents)
      run();
  }

  @Override public synchronized void run()
  {
    isProcessingEvents = true;
    while (!eventQueue.isEmpty())
    {
      Runnable event = eventQueue.remove();
      event.run();
    }
    isProcessingEvents = false;
  }
}

File: src/main/java/com/jackmeng/stl/stl_EventQueue.kt
// Copyright 2023 Jack Meng. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
package com.jackmeng.stl

import java.lang.Runnable
import java.util.LinkedList
import kotlin.jvm.Synchronized

class stl_EventQueue:Runnable
{
	private val eventQueue:LinkedList<Runnable>=LinkedList()
	private var isProcessingEvents:Boolean
	
	init
	{
		isProcessingEvents=false
	}
	
	@Synchronized
	fun add(event:Runnable)
	{
		eventQueue.add(event)
		if (!isProcessingEvents) run()
	}
	
	@Synchronized
	override fun run()
	{
		isProcessingEvents=true
		while (!eventQueue.isEmpty())
		{
			val event=eventQueue.remove()
			event.run()
		}
		isProcessingEvents=false
	}
}